File: samples/cpp/pets/pets.cpp
//-------------------------------------------------------------------
// pets.cpp
//
// A simple example showing an interaction with an
// expert system.
//
// The expert system identifies pets based on sound:
//
//	pets.pro
//
//	pet(dog) :- sound(woof).
//	pet(cat) :- sound(meow).
//	pet(duck) :- sound(quack).
//
// To use it, first a sound/1 fact is asserted, then
// a query is made to pet(X) to determine the type
// of pet.
//
// Compile and link the pets.pro before using it.
// >acmp pets
// >alnk pets pets
//
// In C++ a class is defined that provides the service
// of pet identification to the rest of the application.
// That class wraps the Logic Server as part of its
// implementation.
//
// To test the error handling, rename pets.xpl to something
// else and try again.  You should see the missing .xpl
// file message.
//

#include <iostream>
#include <cstdio>
#include <string>
#include "logicserver.h"

using namespace std;

class petID
{
private:
   LogicServer ls;   // wrap Logic Server in class
public:
	petID();
	~petID();
	void SetSound(string);
	string GetPet();
};

int main()
{
	try
	{
		petID p1;
		string buf;

		p1.SetSound("woof");

		buf = p1.GetPet();
		cout << "pet is a " << buf << '\n';
	}
#ifdef GNU
	// A bug causes segmentation faults when CLSException
	// objects are thrown in Gnu C++, so pointers to the
	// object are thrown instead.  See amzi.h for details.
	catch(LSException *e)
	{
		string errmsg = e->GetMsg();
		cout << "Logic Server Exception " << e->GetRC()
			<< ": " << errmsg << '\n';
		return -1;
	}
#else
	catch(LSException &e)
	{
		string errmsg = e.GetMsg();
		cout << "Logic Server Exception " << e.GetRC()
			<< ": " << errmsg << '\n';
		return -1;
	}
#endif
	return 0;
}

petID::petID()
{
   // Initialize the Logic Server and load the
   // pets.xpl file.

	ls.Init("");
	ls.Load("pets");
}

petID::~petID()
{
	// Close the Logic Server
	ls.Close();
}

void petID::SetSound(string sound)
{
	// put the sound into a string for a Prolog fact
   string buf("sound(" + sound + ")");
	ls.AssertaStr(buf);
}

string petID::GetPet()
{
	TERM t;

	// query for the pet and if successful get
	// the argument to for the caller
   t = ls.ExecStr("pet(X)");
	if ( t )
		return ls.GetStrArg(t, 1);
	else
		return "unknown";
}

File: interfaces/samples/java/pets_callback/Pets.java
/**
 * Pets Local with Extended Predicates
 *
 * This sample uses the tiny pet identification
 * expert system to illustrate extended predicates in Java.
 *
 * This particular example shows how to create a
 * separate class for extended predicates, but
 * the predicates could be just as easily defined
 * in the same class.
 *
 * To test the error handling, rename the file
 * pets.xpl to something else and run the program.
 */

import amzi.ls.*;

class Pets
{
   LogicServer ls;

   public static void main(String args[])
   {
      System.out.println("pets_callback sample");
      try
      {
         Pets p1;
         p1 = new Pets();

         System.out.println("Pet is a " + p1.GetPet());

         p1.close();
      }
      catch (LSException e)
      {
         System.out.println(e.GetMsg());
      }
   }

   public Pets() throws LSException
   {
      // create and initialize a logic server using
      // Init2 to specify small control stack settings
      // for this small application
      ls = new LogicServer();
      ls.Init2("h=1000, l=1000, c=1000, t=1000");

      // create an extended predicate library
      // that initializes its own predicates
      AmziPreds ap = new AmziPreds(ls);

      // load the compiled Prolog expert system
      ls.Load("pets");
   }

   public String GetPet() throws LSException
   {
      long term;

      term = ls.ExecStr("pet(X)");
      if (term != 0)
         return ls.GetStrArg(term, 1);
      else
         return "unknown";
   }

   public void close() throws LSException
   {
      if (ls != null)
         ls.Close();
   }
}
